index.bs

```diff
@@ -1633,8 +1651,8 @@ partial interface MLGraphBuilder {
 Clamp the input tensor element-wise within a range specified by the minimum and maximum values.
 <script type=idl>
 dictionary MLClampOptions {
-  float minValue;
-  float maxValue;
+  MLNumber minValue;
+  MLNumber maxValue;
```

Comment:
FWIW [the Chromium implementation assumes](https://crsrc.org/c/third_party/blink/renderer/modules/ml/webnn/ml_graph_builder.cc;l=530-532;drc=0fb9efef8267777ba45c3ba8e1d934dfe61a0d41) that these values are finite (not unrestricted) floats. In the case of `clamp` it seems reasonable for infinite values to be passed. Barring any pushback, I'd like to propose we change the implementation to allow infinites

That being said, should the `clamp()` function's algorithm throw on `NaN`?

Now that we're allowing unrestricted doubles there's a separate question of whether `minValue` should default to `-Infinity` and `maxValue` should default to `Infinity`... Probably best to leave that to a follow-up :)

Reply:
>  I'd like to propose we change the implementation to allow infinites

Yes, very much so.

> That being said, should the clamp() function's algorithm throw on NaN?

Well NaN propagation is well defined, and there are helper functions specifically intended to check for NaN's during computation ([1](https://www.tensorflow.org/api_docs/python/tf/math/is_nan) [2](https://pytorch.org/docs/stable/generated/torch.isnan.html) [3](https://numpy.org/doc/stable/reference/generated/numpy.isnan.html) [4](https://onnx.ai/onnx/operators/onnx__IsNaN.html)...). So NaN isn't really a foreign concept or illegal concept in ML (then again, maybe it's just because I'm so intimately familiar with dozens of floating-point formats :b). I'd also expect clamp to behave like other libraries. e.g.

```python
x = tensorflow.constant([250], dtype=tensorflow.float32)
y = tensorflow.clip_by_value(x, 13, float('nan'))
print("value:", y)
print("shape:", y.shape)

# Prints:
# value: tf.Tensor([nan], shape=(1,), dtype=float32)
# shape: (1,)
```

> Now that we're allowing unrestricted doubles there's a separate question of whether minValue should default to -Infinity and maxValue should default to Infinity... Probably best to leave that to a follow-up :)

Those are sensible defaults. We just have to be sure that the logic inside Chromium for integer data types like int32 respectively maps +Infinity to INT32_MAX and -Infinity to INT32_MIN. Agreed that it's cleaner to be a separate CR.

Reply:
Cool, I'll add support for infinites and NaN in `clamp` when I [add support for bigint in the bindings code](https://crrev.com/c/5466806)

>  I'd also expect clamp to behave like other libraries. e.g.

Yup, SGTM. We should add WPT conformance tests asserting that we get the expected behavior for inf/NaN (FYI @BruceDai)

Reply:
So for https://pr-preview.s3.amazonaws.com/inexorabletash/webnn/pull/647.html#converttoint it sounds like we'd always want enforceRange = false, clamp = true ? (i.e. N < min &rarr; -&infin;, N > max &rarr; +&infin;)

And for https://pr-preview.s3.amazonaws.com/inexorabletash/webnn/pull/647.html#converttofloat we'd always want restricted = false? (i.e. NaN and +/- &infin; always pass through, N < min &rarr; -&infin;, N > max &rarr; +&infin;)

If that sound right, I can simplify the PR to remove those options.

Reply:
PR updated in 6f75f2b7c041c09cd67fc416849e3c46cb7abbc2 to always clamp.